pyfr/plugins/integrate.py

```diff
@@ -0,0 +1,154 @@
+from pyfr.quadrules import get_quadrule
+
+
+class IntegratePlugin(BasePlugin):
+    name = 'integrate'
+    systems = ['*']
+    formulations = ['dual', 'std']
+
+    def __init__(self, intg, cfgsect, suffix=None):
+        super().__init__(intg, cfgsect, suffix)
+
+        comm, rank, root = get_comm_rank_root()
+        
+        # Underlying elements class
+        self.elementscls = intg.system.elementscls
+
+        # Expressions to integrate
+        c = self.cfg.items_as('constants', float)
+        self.exprs = [self.cfg.getexpr(cfgsect, k, subs=c)
+                      for k in self.cfg.items(cfgsect)
+                      if k.startswith('int-')]
+
+        # Gradient pre-processing
+        self._init_gradients(intg)
+
+        # Save a reference to the physical solution point locations
+        self.plocs = intg.system.ele_ploc_upts
+
+        # Integration parameters
+        self.nsteps = self.cfg.getint(cfgsect, 'nsteps')
+        
+        # The root rank needs to open the output file
+        if rank == root:
+            header = ['t'] + [k for k in self.cfg.items(cfgsect) if \
+                              k.startswith('int-')]
+
+            # Open
+            self.outf = init_csv(self.cfg, cfgsect, ','.join(header))
+            
+        self.elminfo = []
+        # Iterate over the element types
+        for ename, eles in intg.system.ele_map.items():
+            # Locations of each solution point
+            plocupts = eles.ploc_at_np('upts')
+            
+            # Jacobians
+            jacs = 1/eles.rcpdjac_at_np('upts')
+        
+            # Weights
+            rname = self.cfg.get('solver-elements-' + ename, 'soln-pts')
+            wts = get_quadrule(ename, rname, eles.nupts).wts
+            
+            # Save
+            self.elminfo.append((plocupts, jacs, wts))
+
+    def _init_gradients(self, intg):
+        # Determine what gradients, if any, are required
+        self._gradpnames = gradpnames = set()
+        for ex in self.exprs:
+            gradpnames.update(re.findall(r'\bgrad_(.+?)_[xyz]\b', ex))
+
+        # If gradients are required then form the relevant operators
+        if gradpnames:
+            self._gradop, self._rcpjact = [], []
+
+            for eles in intg.system.ele_map.values():
+                self._gradop.append(eles.basis.m4)
+
+                # Get the smats at the solution points
+                smat = eles.smat_at_np('upts').transpose(2, 0, 1, 3)
+
+                # Get |J|^-1 at the solution points
+                rcpdjac = eles.rcpdjac_at_np('upts')
+
+                # Product to give J^-T at the solution points
+                self._rcpjact.append(smat*rcpdjac)
+
+    def _eval_exprs(self, intg):
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        exprs = np.zeros(len(self.exprs))
+        # Iterate over each element type in the simulation
+        for i, (soln, (plocs, jacs, wts)) in \
```

Comment:
Try to avoid \ as they can be a source of subtle issues.  Here I might assign the output of enumerate to a variable first.